File: Lecture 2/MATLAB/distflow.m
% Function to perform power flow calculation using the DistFlow method

function [V,Iteration] = distflow(System)

% Extract the bus data

Number_of_Buses = size(System.Buses,1) ;
P_load = System.Buses.P_load ;
Q_load = System.Buses.Q_load ;

% Extract the branch data

Number_of_Branches = size(System.Branches,1) ;
From_Bus = System.Branches.From_Bus ;
To_Bus = System.Branches.To_Bus ;
R = System.Branches.R ;
X = System.Branches.X ;

% Initialize the calculation variables: arrays of newest and old voltage
% magnitudes, active and reactive power flows at the sending and receiving 
% ends of the branch, as well as iteration variables

V = ones(Number_of_Buses,1) ;
V_old = zeros(Number_of_Buses,1) ;
P = zeros(Number_of_Branches,1) ;
Q = zeros(Number_of_Branches,1) ;
P_rec = zeros(Number_of_Branches,1) ;
Q_rec = zeros(Number_of_Branches,1) ;
Iteration = 0 ;
Tolerance = 1e-6 ;

% Main loop

while any(abs(V-V_old)>=Tolerance)

    % Update the iteration variables

    V_old = V ;
    Iteration = Iteration + 1 ;

    % Backward sweep: calculate active and reactive powers at the sending
    % and receiving ends of each branch

    for i = Number_of_Branches : -1 : 1

        P_rec(i) = P_load(To_Bus(i)) ;
        Q_rec(i) = Q_load(To_Bus(i)) ;

        for j = 1 : Number_of_Branches

            % Check if branch "j" starts from the ending node of branch "i"

            if To_Bus(i) == From_Bus(j)

                P_rec(i) = P_rec(i) + P(j) ;
                Q_rec(i) = Q_rec(i) + Q(j) ;

            end

        end

        % Power flows at the sending end of the branch

        P(i) = P_rec(i) + R(i)*(P_rec(i)^2 + Q_rec(i)^2)/V(To_Bus(i)) ;
        Q(i) = Q_rec(i) + X(i)*(P_rec(i)^2 + Q_rec(i)^2)/V(To_Bus(i)) ;

    end

    % Forward sweep: calculate node voltages

    for i = 1 : Number_of_Branches

        V(To_Bus(i)) = V(From_Bus(i)) - 2*(P(i)*R(i) + Q(i)*X(i)) + ...
            (R(i)^2+X(i)^2)*(P(i)^2+Q(i)^2)/V(From_Bus(i)) ;

    end

end

% Perform voltage correction (for easier programming, up to this point, V
% represents an array of squared voltage magnitudes)

V = sqrt(V) ; 

end

File: Lecture 2/Python/distflow.py
import numpy as np
import pickle


def distflow(system):
    """
    Perform power flow calculations using the DistFlow method.
    """

    # Extract the bus data
    number_of_buses = system['buses'].shape[0]
    p_load = system['buses'][:, 1]
    q_load = system['buses'][:, 2]

    # Extract the branch data
    number_of_branches = system['branches'].shape[0]
    from_bus = system['branches'][:, 1].astype(int)
    to_bus = system['branches'][:, 2].astype(int)
    r = system['branches'][:, 3]
    x = system['branches'][:, 4]

    # Initialize the calculation variables
    V = np.ones(number_of_buses)
    V_old = np.zeros(number_of_buses)
    P = np.zeros(number_of_branches)
    Q = np.zeros(number_of_branches)
    P_rec = np.zeros(number_of_branches)
    Q_rec = np.zeros(number_of_branches)
    tolerance = 1e-6
    iteration = 0

    # Main loop
    while np.any(np.abs(V - V_old) >= tolerance):

        # Update the iteration variables
        V_old = np.copy(V)
        iteration += 1

        # Backward sweep: calculate active and reactive powers at the sending and receiving ends
        for i in range(number_of_branches - 1, -1, -1):
            P_rec[i] = p_load[to_bus[i]]
            Q_rec[i] = q_load[to_bus[i]]

            for j in range(number_of_branches):
                if to_bus[i] == from_bus[j]:
                    P_rec[i] += P[j]
                    Q_rec[i] += Q[j]

            # Power flows at the sending end
            P[i] = P_rec[i] + r[i] * (P_rec[i] ** 2 + Q_rec[i] ** 2) / V[to_bus[i]]
            Q[i] = Q_rec[i] + x[i] * (P_rec[i] ** 2 + Q_rec[i] ** 2) / V[to_bus[i]]

        # Forward sweep: calculate node voltages
        for i in range(number_of_branches):
            V[to_bus[i]] = V[from_bus[i]] - 2 * (P[i] * r[i] + Q[i] * x[i]) + \
                (r[i] ** 2 + x[i] ** 2) * (P[i] ** 2 + Q[i] ** 2) / V[from_bus[i]]

    # Perform voltage correction
    V = np.sqrt(V)

    return V, iteration


# Sample usage
if __name__ == '__main__':
    with open("13 bus system.pkl", "rb") as file:
        system = pickle.load(file)
    V, iteration = distflow(system)
    print("Node voltages:", V)
    print("Number of iterations:", iteration)
